[red-knot] binary arithmetic on instances (#13800)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/binary/instances.md

# N.B. Still a subtype of `A`, even though `A` does not appear directly in the class's `__bases__`
class C(B): ...

# TODO: we currently only understand direct subclasses as subtypes of the superclass.
# We need to iterate through the full MRO rather than just the class's bases;
# if we do, we'll understand `C` as a subtype of `A`, and correctly understand this as being
# `MyString` rather than `str`
reveal_type(A() + C())  # revealed: str
```

## Reflected precedence 2

If the right-hand operand is a subtype of the left-hand operand, but does not
override the reflected method, the left-hand operand's non-reflected method
still takes precedence:

```py
class A:
    def __add__(self, other) -> str: return "foo"
    def __radd__(self, other) -> int: return 42

class B(A): pass

reveal_type(A() + B())  # revealed: str
```

## Only reflected supported

For example, at runtime, `(1).__add__(1.2)` is `NotImplemented`, but
`(1.2).__radd__(1) == 2.2`, meaning that `1 + 1.2` succeeds at runtime
(producing `2.2`).

```py
class A:
    def __sub__(self, other: A) -> A:
        return A()

class B:
    def __rsub__(self, other: A) -> B:
        return B()

> A() - B()

Reviewer: Oops, there should be a `reveal_type` here!

```py
class A:
    def __radd__(self, other) -> A: return self
    def __rsub__(self, other) -> A: return self
    def __rmul__(self, other) -> A: return self
    def __rmatmul__(self, other) -> A: return self
    def __rtruediv__(self, other) -> A: return self
    def __rfloordiv__(self, other) -> A: return self
    def __rmod__(self, other) -> A: return self
    def __rpow__(self, other) -> A: return self
    def __rlshift__(self, other) -> A: return self
    def __rrshift__(self, other) -> A: return self
    def __rand__(self, other) -> A: return self
    def __rxor__(self, other) -> A: return self
    def __ror__(self, other) -> A: return self

class B: pass

reveal_type(B() + A())  # revealed: A
reveal_type(B() - A())  # revealed: A
reveal_type(B() * A())  # revealed: A
reveal_type(B() @ A())  # revealed: A
reveal_type(B() / A())  # revealed: A
reveal_type(B() // A())  # revealed: A
reveal_type(B() % A())  # revealed: A
reveal_type(B() ** A())  # revealed: A
reveal_type(B() << A())  # revealed: A
reveal_type(B() >> A())  # revealed: A
reveal_type(B() & A())  # revealed: A
reveal_type(B() ^ A())  # revealed: A
reveal_type(B() | A())  # revealed: A
```

> ## Return different type

Reviewer: ```suggestion
## Returning a different type
```

The magic methods aren't required to return the type of `self`:

```py
class A:
    def __add__(self, other) -> int: return 1
    def __rsub__(self, other) -> int: return 1

class B:
    pass

reveal_type(A() + B())  # revealed: int
reveal_type(B() - A())  # revealed: int
```

## Non-reflected precedence in general

In general, if the left-hand side defines `__add__` and the right-hand side
defines `__radd__` and the right-hand side is not a subtype of the left-hand
side, `lhs.__add__` will take precedence:

```py
class A:
    def __add__(self, other: B) -> int: return 42

class B:
    def __radd__(self, other: A) -> str: return "foo"

> # C is a subtype of C, but if the two sides are of equal types,
> # the lhs *still* takes precedence
> class C:
>     def __add__(self, other: C) -> int: return 42
>     def __radd__(self, other: C) -> str: return "foo"
> 
> reveal_type(A() + B())  # revealed:  int
> reveal_type(C() + C())  # revealed: int

Reviewer: ```suggestion
reveal_type(A() + B())  # revealed:  int

# Edge case: C is a subtype of C, *but* if the two sides are of *equal* types,
# the lhs *still* takes precedence
class C:
    def __add__(self, other: C) -> int: return 42
    def __radd__(self, other: C) -> str: return "foo"

reveal_type(C() + C())  # revealed: int
```

class B(A):
    def __radd__(self, other) -> MyString: return MyString()

reveal_type(A() + B())  # revealed: MyString

# N.B. Still a subtype of `A`, even though `A` does not appear directly in the class's `__bases__`
class C(B): ...

# TODO: we currently only understand direct subclasses as subtypes of the superclass.
# We need to iterate through the full MRO rather than just the class's bases;
# if we do, we'll understand `C` as a subtype of `A`, and correctly understand this as being
# `MyString` rather than `str`
reveal_type(A() + C())  # revealed: str
```

## Reflected precedence 2

If the right-hand operand is a subtype of the left-hand operand, but does not
override the reflected method, the left-hand operand's non-reflected method
still takes precedence:

```py
class A:
    def __add__(self, other) -> str: return "foo"
    def __radd__(self, other) -> int: return 42

class B(A): pass

reveal_type(A() + B())  # revealed: str
```

## Only reflected supported

For example, at runtime, `(1).__add__(1.2)` is `NotImplemented`, but
`(1.2).__radd__(1) == 2.2`, meaning that `1 + 1.2` succeeds at runtime
(producing `2.2`). The runtime tries the second one only if the first one

> returns `NotImplemented` to signal failure .

Reviewer: ```suggestion
returns `NotImplemented` to signal failure.
```

Typeshed and other stubs annotate dunder-method calls that would return
`NotImplemented` as being "illegal" calls. `int.__add__` is annotated as only
"accepting" `int`s, even though it strictly-speaking "accepts" any other object

> with raising an error -- it will simply return `NotImplemented`, allowing the
> runtime to try the `__radd__` method of the right-hand operand as well.

Reviewer: ```suggestion
without raising an exception -- it will simply return `NotImplemented`,
allowing the runtime to try the `__radd__` method of the right-hand operand
as well.
```

```py
class A:
    def __sub__(self, other: A) -> A:
        return A()

class B:
    def __rsub__(self, other: A) -> B:
        return B()

# TODO: this should be `B` (the return annotation of `B.__rsub__`),
# because `A.__sub__` is annotated as only accepting `A`,
# but `B.__rsub__` will accept `A`.
reveal_type(A() - B())  # revealed: A
```

## Callable instances as dunders

Believe it or not, this is supported at runtime:

```py
class A:
    def __call__(self, other) -> int:
        return 42

class B:
    __add__ = A()

reveal_type(B() + B())  # revealed: int
```

## Integration test: numbers from typeshed

```py
reveal_type(3j + 3.14)  # revealed: complex
reveal_type(4.2 + 42)  # revealed: float
reveal_type(3j + 3)  # revealed: complex

> # TODO should be complex, need to check arg type and fall back
> reveal_type(3.14 + 3j)  # revealed: float
> 
> # TODO should be float, need to check arg type and fall back
> reveal_type(42 + 4.2)  # revealed: int
> 
> # TODO should be complex, need to check arg type and fall back
> reveal_type(3 + 3j)  # revealed: int

Reviewer: ```suggestion
# TODO should be complex, need to check arg type and fall back to `rhs.__radd__`
reveal_type(3.14 + 3j)  # revealed: float

# TODO should be float, need to check arg type and fall back to `rhs.__radd__`
reveal_type(42 + 4.2)  # revealed: int

# TODO should be complex, need to check arg type and fall back to `rhs.__radd__`
reveal_type(3 + 3j)  # revealed: int
```

def returns_int() -> int:
    return 42

def returns_bool() -> bool:
    return True

x = returns_bool()
y = returns_int()

reveal_type(x + y)  # revealed: int
reveal_type(4.2 + x)  # revealed: float

> # TODO should be float, need to check arg type and fall back

Reviewer: ```suggestion
# TODO should be float, need to check arg type and fall back to `rhs.__radd__`
```

# TODO overloads
reveal_type(literal_string_instance + A())  # revealed: @Todo
```

## Operations involving instances of classes inheriting from `Any`

`Any` and `Unknown` represent a set of possible runtime objects, wherein the
bounds of the set are unknown. Whether the left-hand operand's dunder or the
right-hand operand's reflected dunder depends on whether the right-hand operand
is an instance of a class that is a subclass of the left-hand operand's class
and overrides the reflected dunder. In the following example, because of the
unknowable nature of `Any`/`Unknown`, we must consider both possibilities:
`Any`/`Unknown` might resolve to an unknown third class that inherits from `X`
and overrides `__radd__`; but it also might not. Thus, the correct answer here
for the `reveal_type` is `int | Unknown`.

```py
from does_not_exist import Foo  # error: [unresolved-import]

reveal_type(Foo)  # revealed: Unknown

class X:
    def __add__(self, other: object) -> int:
        return 42

class Y(Foo):
    pass

# TODO: Should be `int | Unknown`; see above discussion.
reveal_type(X() + Y())  # revealed: int
```

## Unsupported

> ### Dunder on instance

Reviewer: ```suggestion
### Dunder as instance attribute
```

> The magic method must be on the class, not just on the instance:

Reviewer: ```suggestion
The magic method must exist on the class, not just on the instance:
```
